File: src/py_pubsub/py_pubsub/topics_practice.py
import rclpy
from rclpy.node import Node
from geometry_msgs.msg import Twist
from turtlesim.msg import Pose

class MotionController(Node):
    
    def __init__(self):
        super().__init__('motion_controller')
        
        self.publisher_ = self.create_publisher(Twist, '/turtle1/cmd_vel', 10)
        self.subscription_ = self.create_subscription(Pose, '/turtle1/pose', self.pose_callback, 10)
        self.timer_ = self.create_timer(0.1, self.timer_callback)
        
        self.pose_ = Pose()
        self.linear_velocity = 0.5
        self.angular_velocity = 1.0
        self.init_pos_x = 5.5
        self.init_pos_y = 5.5
        self.in_bounds = True
        self.counter = 5 

    def pose_callback(self, msg):
        self.pose_.x = msg.x
        self.pose_.y = msg.y
        
        MIN_BORDER = 2.0
        MAX_BORDER = 8.0

        if MIN_BORDER < self.pose_.x < MAX_BORDER and MIN_BORDER < self.pose_.y < MAX_BORDER:
            self.in_bounds = True
        else:
            self.in_bounds = False

    def timer_callback(self):
        if self.counter > 0:
            self.get_logger().info(f'Counter: {self.counter}')
            self.counter -= 1
            return
        
        twist_msg = Twist()
        
        if self.in_bounds:
            twist_msg.linear.x = self.linear_velocity
            twist_msg.angular.z = self.angular_velocity
            
            self.linear_velocity += 0.01
            self.angular_velocity = 1.0 / self.linear_velocity
            
            self.get_logger().info('Drawing spiral')
        else:
            twist_msg.linear.x = self.linear_velocity
            twist_msg.angular.z = 0.0
            self.get_logger().info('Going straight')
        
        self.publisher_.publish(twist_msg)

def main(args=None):
    rclpy.init(args=args)
    motion_controller = MotionController()
    rclpy.spin(motion_controller)
    motion_controller.destroy_node()
    rclpy.shutdown()

if __name__ == '__main__':
    main()

File: src/cpp_pubsub/src/topics_practice.cpp
#include <chrono>
#include <memory>
#include "rclcpp/rclcpp.hpp"
#include "geometry_msgs/msg/twist.hpp"
#include "turtlesim/msg/pose.hpp"

using namespace std::chrono_literals;

class MotionController : public rclcpp::Node {
public:
    MotionController() : Node("motion_controller"), counter_(5), linear_velocity_(0.5), angular_velocity_(1.0), in_bounds_(true) {
        publisher_ = this->create_publisher<geometry_msgs::msg::Twist>("/turtle1/cmd_vel", 10);
        subscription_ = this->create_subscription<turtlesim::msg::Pose>(
            "/turtle1/pose", 10, std::bind(&MotionController::pose_callback, this, std::placeholders::_1));
        timer_ = this->create_wall_timer(100ms, std::bind(&MotionController::timer_callback, this));
    }

private:
    void pose_callback(const turtlesim::msg::Pose::SharedPtr msg) {
        pose_.x = msg->x;
        pose_.y = msg->y;

        const double MIN_BORDER = 2.0;
        const double MAX_BORDER = 8.0;

        if (MIN_BORDER < pose_.x && pose_.x < MAX_BORDER && MIN_BORDER < pose_.y && pose_.y < MAX_BORDER) {
            in_bounds_ = true;
        } else {
            in_bounds_ = false;
        }
    }

    void timer_callback() {
        if (counter_ > 0) {
            RCLCPP_INFO(this->get_logger(), "Counter: %d", counter_);
            --counter_;
            return;
        }

        auto twist_msg = geometry_msgs::msg::Twist();

        if (in_bounds_) {
            twist_msg.linear.x = linear_velocity_;
            twist_msg.angular.z = angular_velocity_;

            linear_velocity_ += 0.01;
            angular_velocity_ = 1.0 / linear_velocity_;

            RCLCPP_INFO(this->get_logger(), "Drawing spiral");
        } else {
            twist_msg.linear.x = linear_velocity_;
            twist_msg.angular.z = 0.0;

            RCLCPP_INFO(this->get_logger(), "Going straight");
        }

        publisher_->publish(twist_msg);
    }

    rclcpp::Publisher<geometry_msgs::msg::Twist>::SharedPtr publisher_;
    rclcpp::Subscription<turtlesim::msg::Pose>::SharedPtr subscription_;
    rclcpp::TimerBase::SharedPtr timer_;
    turtlesim::msg::Pose pose_;
    int counter_;
    double linear_velocity_;
    double angular_velocity_;
    bool in_bounds_;
};

int main(int argc, char *argv[]) {
    rclcpp::init(argc, argv);
    rclcpp::spin(std::make_shared<MotionController>());
    rclcpp::shutdown();
    return 0;
}
